### PR TITLE
Optimize flow of execution for test reconciliation

### PR DESCRIPTION
**Pull request by codeblooded**
The previous flow of execution for the (*LoadTestReconciler).Reconcile method included pre-emptively fetching resources and making unnecessary calls to account for defaults. Aside from the unnecessary requests, the flow was hard to follow.

This change reorders the method, removing unnecessary calls and switching to fetch resources lazily.

### CODE REVIEW

File: controllers/loadtest_controller.go [hunk 1]
```diff
@@ -56,64 +57,44 @@ type LoadTestReconciler struct {
...
-		log.Error(err, "failed to list pods", "namespace", req.Namespace)
-		return ctrl.Result{Requeue: true}, err
-	}
-
-	loadtests := new(grpcv1.LoadTestList)
-	if err = r.List(ctx, loadtests); err != nil {
-		log.Error(err, "failed to list loadtests")
-		return ctrl.Result{Requeue: true}, err
-	}
-
```
Comment (wanlin31) [inline controllers/loadtest_controller.go:86]: I like that you removed this part, I used to think all the time why we have fetch all the LoadTest and  fetch the one that actually triggered the reconciliation later. Make more scene and more efficient this way.

File: controllers/loadtest_controller.go [hunk 2]
```diff
@@ -56,64 +57,44 @@ type LoadTestReconciler struct {
...
 	loadtest := new(grpcv1.LoadTest)
 	if err = r.Get(ctx, req.NamespacedName, loadtest); err != nil {
 		log.Error(err, "failed to get loadtest", "name", req.NamespacedName)
-
-		// do not requeue, the load test may have been deleted
```
Comment (wanlin31) [inline controllers/loadtest_controller.go:91]: why remove the comments? I found them pretty useful :)

File: controllers/loadtest_controller.go [hunk 3]
```diff
@@ -56,64 +57,44 @@ type LoadTestReconciler struct {
...
+	if err = r.Defaults.SetLoadTestDefaults(loadtestWithDefaults); err != nil {
+		log.Error(err, "failed to clone test with defaults")
 		return ctrl.Result{}, err
 	}
-	if err = r.Update(ctx, loadtest); err != nil {
-		log.Error(err, "failed to update loadtest with defaults")
+	if !reflect.DeepEqual(loadtest, loadtestWithDefaults) {
+		if err = r.Update(ctx, loadtestWithDefaults); err != nil {
+			log.Error(err, "failed to update test with defaults")
+			return ctrl.Result{}, err
```
Comment (wanlin31) [inline controllers/loadtest_controller.go:84]: Little confused about the update function, I used to think it happens like this: every time we get a triggering LoadTest -> we made sure that it follows the default -> we update the triggering LoadTest. I understand you make the change to make it a lazy update, but say if we do found a difference (that is line 82), why are we updating the `loadtestWithDefaults` instead of `loadtest`?

Comment (codeblooded) [reply controllers/loadtest_controller.go:84]: `loadtestWithDefaults` is equivalent to `loadtest`, except it adds defaults. It's name, namespace, and GVK (group, version, kind) information is identical. If it looks like a duck... walks like a duck... quacks like a duck... it is a duck. To Kubernetes, this is enough for it to know which object we are talking about (even if it  is not the same object in memory). It will be treated the same. We don't want to update with loadtest, because that would do nothing. It has no defaults. They both are kept around, b/c line 81 ensures that we do not update every single time we reconcile (only when we need to set defaults).

File: controllers/loadtest_controller.go (hunk 2 above)
Comment (codeblooded) [reply controllers/loadtest_controller.go:91]: Good to know. I did not realize they were useful. I thought they may be redundant since we had a comment every time we reconciled. I can see it can be useful in some contexts.

**Comment by codeblooded** [comment]
Moving to draft, because I will rename some of these variables and improve it a little. Will move back to ready when fixed and re-request review. I think this will aid with clarity.

File: controllers/loadtest_controller.go (hunk 3 above)
Comment (wanlin31) [reply controllers/loadtest_controller.go]: " To Kubernetes, this is enough for it to know which object we are talking about (even if it is not the same object in memory)" Got it ! Thx. :)

**Review by wanlin31 — APPROVED** [review]

**Comment by codeblooded** [comment]
@wanlin31 I added back the comment that you liked. I think this should be good to go. @paulosjca do you want to review?

**Review by paulosjca — APPROVED** [review]
